File: sdk/python/tempmail_sdk/providers/tempfastmail.py
"""
TempFastMail 渠道 -- https://tempfastmail.com
"""

from __future__ import annotations

from typing import Any, Dict, List
from urllib.parse import quote

from ..normalize import normalize_email
from ..types import Email, EmailInfo
from .. import http

CHANNEL = "tempfastmail"
BASE_URL = "https://tempfastmail.com"
HEADERS = {"Accept": "application/json", "User-Agent": "Mozilla/5.0"}


def _flatten(raw: Dict[str, Any], recipient: str) -> Dict[str, Any]:
    return {
        "id": raw.get("uuid"),
        "from": raw.get("from"),
        "to": raw.get("real_to") or recipient,
        "subject": raw.get("subject"),
        "html": raw.get("html"),
        "date": raw.get("received_at"),
        "attachments": raw.get("attachments"),
    }


def generate_email() -> EmailInfo:
    response = http.post(f"{BASE_URL}/api/email-box", headers=HEADERS)
    response.raise_for_status()
    data = response.json()
    email = str(data.get("email") or "").strip()
    uuid = str(data.get("uuid") or "").strip()
    if not email or "@" not in email or not uuid:
        raise ValueError("tempfastmail: invalid create response")
    return EmailInfo(channel=CHANNEL, email=email, token=uuid)


def get_emails(token: str, email: str) -> List[Email]:
    uuid = str(token or "").strip()
    address = str(email or "").strip()
    if not uuid:
        raise ValueError("tempfastmail: empty token")
    if not address:
        raise ValueError("tempfastmail: empty email")

    response = http.get(
        f"{BASE_URL}/api/email-box/{quote(uuid, safe='')}/emails", headers=HEADERS
    )
    response.raise_for_status()
    rows = response.json()
    if not isinstance(rows, list):
        return []

    out: List[Email] = []
    for row in rows:
        raw = row if isinstance(row, dict) else {}
        message_id = str(raw.get("uuid") or "").strip()
        if message_id:
            detail_response = http.get(
                f"{BASE_URL}/api/email-box/{quote(uuid, safe='')}/email/{quote(message_id, safe='')}",
                headers=HEADERS,
            )
            if detail_response.ok:
                detail = detail_response.json()
                if isinstance(detail, dict):
                    raw = detail
        out.append(normalize_email(_flatten(raw, address), address))
    return out

File: sdk/ruby/lib/tempmail_sdk/providers/tempfastmail.rb
# frozen_string_literal: true

require "uri"

module TempmailSdk
  module Providers
    # TempFastMail 渠道 — https://tempfastmail.com
    module Tempfastmail
      CHANNEL = "tempfastmail"
      BASE_URL = "https://tempfastmail.com"
      HEADERS = { "Accept" => "application/json", "User-Agent" => "Mozilla/5.0" }.freeze

      module_function

      def flatten(raw, recipient)
        {
          "id" => raw["uuid"],
          "from" => raw["from"],
          "to" => raw["real_to"] || recipient,
          "subject" => raw["subject"],
          "html" => raw["html"],
          "date" => raw["received_at"],
          "attachments" => raw["attachments"]
        }
      end

      # 创建临时邮箱
      # @return [EmailInfo]
      def generate_email
        resp = Http.post("#{BASE_URL}/api/email-box", headers: HEADERS)
        resp.raise_for_status
        data = resp.json
        email = (data["email"] || "").strip
        uuid = (data["uuid"] || "").strip
        raise "tempfastmail: 创建邮箱响应无效" if email.empty? || !email.include?("@") || uuid.empty?

        EmailInfo.new(channel: CHANNEL, email: email, token: uuid)
      end

      # 获取邮件列表
      # @param token [String] UUID
      # @param email [String]
      # @return [Array<Email>]
      def get_emails(token, email)
        uuid = (token || "").strip
        address = (email || "").strip
        raise "tempfastmail: token 为空" if uuid.empty?
        raise "tempfastmail: 邮箱地址为空" if address.empty?

        resp = Http.get("#{BASE_URL}/api/email-box/#{URI.encode_www_form_component(uuid)}/emails",
                        headers: HEADERS)
        resp.raise_for_status
        rows = resp.json
        return [] unless rows.is_a?(Array)

        rows.filter_map do |row|
          raw = row.is_a?(Hash) ? row : {}
          message_id = (raw["uuid"] || "").strip
          unless message_id.empty?
            begin
              dr = Http.get("#{BASE_URL}/api/email-box/#{URI.encode_www_form_component(uuid)}/email/#{URI.encode_www_form_component(message_id)}",
                            headers: HEADERS)
              if dr.ok?
                detail = dr.json
                raw = detail if detail.is_a?(Hash)
              end
            rescue StandardError
              # 忽略
            end
          end
          Normalize.normalize_email(flatten(raw, address), address)
        end
      end
    end
  end
end
